### DISCUSSION TITLE
packageRules cannot combine both matchUpdateTypes and rangeStrategy

### DISCUSSION THREAD

**Discussion by coderbyheart** [renovatebot/renovate#10165] [Request Help] [answered]
19 days ago Renovate [opened an issue](https://github.com/NordicSemiconductor/asset-tracker-cloud-code-style-js/issues/81), stating that the configuration is invalid (it used to work however):

> Error type: The renovate configuration file contains some invalid settings
> Message: `packageRules[0]: packageRules cannot combine both matchUpdateTypes and rangeStrategy. Rule: {"automerge":true,"rangeStrategy":"bump","matchUpdateTypes":["minor","pin","digest"],"matchDepTypes":["dependencies","devDependencies","optionalDependencies","peerDependencies"]}, packageRules[1]: packageRules cannot combine both matchUpdateTypes and rangeStrategy. Rule:`

However, after reviewing the documentation, I don't see why I can't and should not longer be able to specify this configuration:

```json
{
  "extends": ["config:base"],
  "ignorePresets": [":prHourlyLimit2", ":prConcurrentLimit20"],
  "packageRules": [
    {
      "matchUpdateTypes": ["minor", "pin", "digest"],
      "automerge": true,
      "rangeStrategy": "bump",
      "depTypeList": [
        "dependencies",
        "devDependencies",
        "optionalDependencies",
        "peerDependencies"
      ]
    },
    {
      "matchUpdateTypes": ["patch", "lockFileMaintenance"],
      "automerge": true,
      "automergeType": "branch",
      "rangeStrategy": "bump",
      "depTypeList": [
        "dependencies",
        "devDependencies",
        "optionalDependencies",
        "peerDependencies"
      ]
    }
  ]
}
```

**Comment by rarkins** [comment, marked as answer]
`rangeStrategy` is used *before* update types are assigned. Therefore you can't assign a `rangeStrategy` with a rule that matches update types, because it will be assigned too late. Try splitting your rules:

```
{
  "extends": ["config:base"],
  "ignorePresets": [":prHourlyLimit2", ":prConcurrentLimit20"],
  "packageRules": [
    {
      "rangeStrategy": "bump",
      "depTypeList": [
        "dependencies",
        "devDependencies",
        "optionalDependencies",
        "peerDependencies"
      ]
    },
    {
      "matchUpdateTypes": ["minor", "pin", "digest"],
      "automerge": true,
      "depTypeList": [
        "dependencies",
        "devDependencies",
        "optionalDependencies",
        "peerDependencies"
      ]
    },
    {
      "rangeStrategy": "bump",
      "depTypeList": [
        "dependencies",
        "devDependencies",
        "optionalDependencies",
        "peerDependencies"
      ]
    },
    {
      "matchUpdateTypes": ["patch", "lockFileMaintenance"],
      "automerge": true,
      "automergeType": "branch",
      "depTypeList": [
        "dependencies",
        "devDependencies",
        "optionalDependencies",
        "peerDependencies"
      ]
    }
  ]
}
```

**Reply by coderbyheart** [reply]
Thanks!

**Reply by rarkins** [reply]
And no thanks at all for this:

![image](https://user-images.githubusercontent.com/6311784/119655067-6f967500-be29-11eb-8d52-26963f24a304.png)

You seem to have posted that complaint the same minute you posted the discussion here. Meanwhile you had a response in 6 minutes before I'd seen the tweet. How about being supportive of open source and free tools instead?

**Reply by rarkins** [reply]
Correction: You apparently tweeted the complaint 12 minutes *before* you posted the discussion here.
